File: go-backend-interview-prep/06_concurrency/07_waitgroups/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func worker(id int, wg *sync.WaitGroup) {
	defer wg.Done() // always defer
	fmt.Printf("  worker %d: started\n", id)
	time.Sleep(50 * time.Millisecond)
	fmt.Printf("  worker %d: done\n", id)
}

func main() {
	// --- Example 1: basic WaitGroup ---
	fmt.Println("=== Basic WaitGroup ===")
	var wg sync.WaitGroup

	for i := 1; i <= 5; i++ {
		wg.Add(1) // Add BEFORE go
		go worker(i, &wg)
	}
	wg.Wait() // blocks until all Done()
	fmt.Println("all workers finished")

	// --- Example 2: WaitGroup + channel for results ---
	fmt.Println("\n=== WaitGroup + channel for results ===")
	results := make(chan int, 5)

	var wg2 sync.WaitGroup
	for i := 1; i <= 5; i++ {
		wg2.Add(1)
		go func(n int) {
			defer wg2.Done()
			results <- n * n
		}(i)
	}

	// Close channel after all goroutines finish
	go func() {
		wg2.Wait()
		close(results)
	}()

	for r := range results {
		fmt.Printf("  result: %d\n", r)
	}

	// --- Example 3: Add(n) in bulk ---
	fmt.Println("\n=== Add(n) in bulk ===")
	var wg3 sync.WaitGroup
	n := 3
	wg3.Add(n) // add all at once

	for i := 0; i < n; i++ {
		go func(id int) {
			defer wg3.Done()
			fmt.Printf("  task %d complete\n", id)
		}(i)
	}
	wg3.Wait()
	fmt.Println("all tasks done")
}

File: go-backend-interview-prep/06_concurrency/07_waitgroups/main.py
"""WaitGroups -- Python equivalent of the Go example.

Python equivalent: thread.join() or concurrent.futures.
Thread.join() blocks until the thread finishes, like wg.Wait().
"""

import threading
import queue
import time
from concurrent.futures import ThreadPoolExecutor, wait


def worker(wid: int) -> None:
    print(f"  worker {wid}: started")
    time.sleep(0.05)
    print(f"  worker {wid}: done")


def main() -> None:
    # --- Example 1: basic join (WaitGroup equivalent) ---
    print("=== Basic join ===")
    threads = []
    for i in range(1, 6):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()  # blocks until thread finishes
    print("all workers finished")

    # --- Example 2: threads + queue for results ---
    print("\n=== Threads + queue for results ===")
    results: queue.Queue[int] = queue.Queue()
    threads = []

    for i in range(1, 6):
        def task(n):
            results.put(n * n)
        t = threading.Thread(target=task, args=(i,))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    while not results.empty():
        print(f"  result: {results.get()}")

    # --- Example 3: ThreadPoolExecutor (higher-level) ---
    print("\n=== ThreadPoolExecutor ===")
    with ThreadPoolExecutor(max_workers=3) as executor:
        futures = [
            executor.submit(lambda tid: print(f"  task {tid} complete"), i)
            for i in range(3)
        ]
        wait(futures)  # blocks until all complete
    print("all tasks done")


if __name__ == "__main__":
    main()
